esmvalcore/preprocessor/_area.py

```diff
@@ -177,22 +177,44 @@ def tile_grid_areas(cube, fx_files):
             fx_cube = iris.load_cube(fx_file)
 
             grid_areas = fx_cube.core_data()
-            if cube.ndim == 4 and grid_areas.ndim == 2:
-                grid_areas = da.tile(grid_areas,
-                                     [cube.shape[0], cube.shape[1], 1, 1])
-            elif cube.ndim == 4 and grid_areas.ndim == 3:
-                grid_areas = da.tile(grid_areas, [cube.shape[0], 1, 1, 1])
-            elif cube.ndim == 3 and grid_areas.ndim == 2:
-                grid_areas = da.tile(grid_areas, [cube.shape[0], 1, 1])
-            else:
-                raise ValueError('Grid and dataset number of dimensions not '
-                                 'recognised: {} and {}.'
-                                 ''.format(cube.ndim, grid_areas.ndim))
+    return grid_areas
+
+
+def tile_grid_areas(cube, grid_areas):
+    """
+    Tile the grid area data to match the dataset cube.
+
+    Parameters
+    ----------
+    cube: iris.cube.Cube
+        input cube.
+    fx_files: dict
+        dictionary of field:filename for the fx_files
+
+    Returns
+    -------
+    iris.cube.Cube
+        Freshly tiled grid areas cube.
+    """
+    if cube.ndim == grid_areas.ndim:
+        return grid_areas
+    if cube.ndim == 4 and grid_areas.ndim == 2:
+        tiled = da.array([grid_areas for i in range(cube.shape[1])])
+        grid_areas = da.stack(tiled, axis=0)
+        tiled = da.array([grid_areas for i in range(cube.shape[0])])
+        grid_areas = da.stack(tiled, axis=0).compute()
+    elif (cube.ndim, grid_areas.ndim) in [(4, 3), (3, 2)]:
+        tiled = da.array([grid_areas for i in range(cube.shape[0])])
+        grid_areas = da.stack(tiled, axis=0).compute()
+    else:
+        raise ValueError('Grid and dataset number of dimensions not '
+                         'recognised: {} and {}.'
+                         ''.format(cube.ndim, grid_areas.ndim))
     return grid_areas
 
 
 # get the area average
-def area_statistics(cube, operator, fx_files=None):
+def area_statistics(cube, operator, fx_files=None, calculate_grid=False):
```

Comment:
I think it would be nicer to define this function as
```python
def area_statistics(cube, operator, weights=None):
```
and `weights` can be either the path to an fx file, the string `'calculate'` or `None` for unweighted statistics.